### PR TITLE
make_string for arguments, which have `operator<<`

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: Michał Szołucha <mszolucha@nvidia.com>

#### Why we need this PR?
*Pick one*
- It adds new feature: creating a function-like string for bunch of arguments, which have only `operator<<`

We need it, because with it we can do like:
```
vec<3> v={1,2,3}
DALI_ENFORCE(something, make_string("Failed because of: ", v));
```
Since we can't do `std::to_string(v)` and polluting `std` with custom overload of `to_string` seems like ugly solution.

### CODE REVIEW

File: dali/util/make_string.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+template <class T>
+std::string make_string(std::stringstream &ss, const T &val) {
+  ss << val;
+  return ss.str();
+}
+
+
+template <class T, class... Args>
+std::string make_string(std::stringstream &ss, const T &val, const Args &... args) {
+  ss << val;
```
Comment (awolant) [inline dali/util/make_string.h]: Maybe consider some kind of spacing between the arguments? Right now it might be hard to read the resulting string.

Comment (JanuszL) [reply dali/util/make_string.h]: :+1:

Comment (szalpal) [reply dali/util/make_string.h]: I omitted the spacing intentionally, as that's string concatenation, not printing itself. I can imagine the situation, that it can be used in case, where such spacings will be no-go.

Comment (jantonguirao) [reply dali/util/make_string.h]: Actually, I agree with @awolant. The common use case of this would be the one with spacing between the arguments. You can either make it configurable (passing a delimiter argument) or make a choice by default (e.g. `,` or ` `)

Comment (jantonguirao) [reply dali/util/make_string.h]: Actually, I agree with @awolant here. I think the most common use case of this will be displaying a bunch of values separated by spaces. I'd rather do `make_string(str1, 2, 3)` than `make_string(str1, " ", 2, " ", 3)`. Consider making this configurable by passing a delimiter (perhaps with a default value of `" "` for the delimiter)

Comment (klecki) [reply dali/util/make_string.h]: That function turned out to be a `concat` ∘ `to_string`. And it will most probably be now used for most of the DALI_ENFORCE stuff to easily print messages, in that case we can do with `make_string("Images don't have matching number of channels, expected ", x, " found ", y, ".");`

Comment (klecki) [reply dali/util/make_string.h]: But I'am not sure if it's the best solution that we can have.

Comment (JanuszL) [reply dali/util/make_string.h]: If `vec<3>` needs to be printed then the delimiter is more than welcome. But in case of `make_string("Images don't have matching number of channels, expected ", x, " found ", y, ".");` rather not.
How about two variants:
* `make_string` without delimiter
* `make_string_delim` with delimeter
?

Comment (szalpal) [reply dali/util/make_string.h]: Done

**Comment by szalpal** [comment]
> I have a feeling that in the current state of this API most of the calls will look like:
> 
> ```
> vec<3> v={1,2,3}
> DALI_ENFORCE(something, make_string_delim(" ", "Failed on:", v[0], "because of:", v[1]));
> ```
> 
> Maybe consider the `make_string` to be wrap for `make_string_delim` with `delim=" "`. With this you can do:
> 
> ```
> vec<3> v={1,2,3}
> DALI_ENFORCE(something, make_string("Failed on because of:", v));
> ```
> 
> and get the same thing as before with one less character and less code to maintain. When you need string to be without spaces you can pass `delim=""`.

Done

**Comment by szalpal** [comment]
!build

File: dali/util/make_string_test.cc [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <gtest/gtest.h>
+#include "dali/core/geom/vec.h"
+#include "dali/util/make_string.h"
+
+namespace dali {
+
+TEST(MakeStringTest, default_delimiter) {
+  auto str = make_string("jeden", 2, 3);
+  ASSERT_EQ(str, "jeden 23");
```
Comment (jantonguirao) [inline dali/util/make_string_test.cc]: this should be "jeden 2 3"

Comment (jantonguirao) [reply dali/util/make_string_test.cc]: Also, EXPECT_EQ is more appropriate here IMHO

**Comment by szalpal** [comment]
!build

File: dali/util/make_string_test.cc (hunk 2 above)
Comment (szalpal) [reply dali/util/make_string_test.cc]: Done

**Comment by szalpal** [comment]
!build

File: dali/util/make_string_test.cc [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+TEST(MakeStringTest, default_delimiter) {
+  auto str = make_string("jeden", 2, 3);
+  ASSERT_EQ(str, "jeden 2 3");
+}
+
+TEST(MakeStringTest, custom_delimiter) {
+  auto str = make_string_delim("]:->", "jeden", 2, 3);
+  ASSERT_EQ(str, "jeden]:->2]:->3");
+}
+
```
Comment (klecki) [inline dali/util/make_string_test.cc]: Please add some tests for boundary conditions:
* no arguments
* one argument
* only delimiter
* delimiter + one argument.

File: dali/util/make_string.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+template <class T>
+std::string make_string_delim(const std::string &delimiter, std::stringstream &ss, const T &val) {
+  ss << val;
+  return ss.str();
+}
+
+
+template <class T, class... Args>
+std::string make_string_delim(const std::string &delimiter, std::stringstream &ss, const T &val,
```
Comment (mzient) [inline dali/util/make_string.h]: This is bad design. It entangles two different concepts: print_delim and make_string_delim.
There should be a function print_delim taking a generic `std::ostream &` and make_string_delim written as:
```
template <typename Delimiter, typename... Args>
std::string make_string_delim(const Delimiter &delim, Args&& ...args) {
  std::stringstream ss;
  print_delim(ss, delim, std::forward<Args>(args)...);
  return ss.str();
}
```

File: include/dali/core/format.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// limitations under the License.#ifndef DALI_MAKE_STRING_H
+
+#ifndef DALI_CORE_FORMAT_H_
+#define DALI_CORE_FORMAT_H_
+
+#include <string>
+
+namespace dali {
+
+template<class Delimiter>
```
Comment (jantonguirao) [inline include/dali/core/format.h]: typename

File: include/dali/core/format.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+void print_delim(std::ostream&os, const Delimiter &delimiter) {
+  // No-op
+}
+
+
+template <class Delimiter, class T>
```
Comment (jantonguirao) [inline include/dali/core/format.h]: and here

File: include/dali/core/format.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+void print_delim(std::ostream &os, const Delimiter &delimiter, const T& val) {
+  os << val;
+}
+
+/**
+ * Populates stream with given arguments, as long as they have
+ * `operator<<` defined for stream operation.
+ */
+template <class Delimiter, class T, class... Args>
```
Comment (jantonguirao) [inline include/dali/core/format.h]: and here

File: include/dali/core/format.h [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+ * Creates std::string from arguments, as long as every element has `operator<<`
+ * defined for stream operation.
+ *
+ * If there's no `operator<<`, compiler will fire an error, saying:
+ * " no match for ‘operator<<’ (operand types are ‘std::stringstream’ and ‘<your-type-here>’) "
+ *
+ * @param delimiter String, which will separate arguments in the final string
+ * @return Concatenated std::string
+*/
+template <class Delimiter, class... Args>
```
Comment (jantonguirao) [inline include/dali/core/format.h]: and here (won't repeat for occurrences below)

File: include/dali/core/format.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  print_delim(os, delimiter, args...);
+}
+
+
+/**
+ * Creates std::string from arguments, as long as every element has `operator<<`
+ * defined for stream operation.
+ *
+ * If there's no `operator<<`, compiler will fire an error, saying:
+ * " no match for ‘operator<<’ (operand types are ‘std::stringstream’ and ‘<your-type-here>’) "
```
Comment (mzient) [inline include/dali/core/format.h]: Don't say what the compiler will do - it's sufficient to say that the code won't compile.

File: include/dali/core/format.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+
+template <class Delimiter, class T>
+void print_delim(std::ostream &os, const Delimiter &delimiter, const T& val) {
+  os << val;
+}
+
+/**
+ * Populates stream with given arguments, as long as they have
```
Comment (mzient) [inline include/dali/core/format.h]:
```suggestion
 * @brief Populates stream with given arguments, as long as they have
```

File: include/dali/core/format.h [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+template <class Delimiter, class... Args>
+std::string make_string_delim(const Delimiter &delimiter, const Args &... args) {
+  std::stringstream ss;
+  print_delim(ss, delimiter, args...);
+  return ss.str();
+}
+
+
+/**
+ * In case there are no Args to operate on, return empty string.
```
Comment (mzient) [inline include/dali/core/format.h]: When looked at in doxygen, this is out of context. This doesn't even have to be a Doxygen comment - and if you insist on it being one, give the reader some context, e.g.
```suggestion
 * This overload handles the edge case when no arguments are given and returns an empty string.
```

File: include/dali/core/format.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+ */
+template <class Delimiter>
+std::string make_string_delim(const Delimiter &delimiter) {
```
Comment (mzient) [inline include/dali/core/format.h]: Remove the argument name to avoid potential warnings `unused argument` or sth.

File: include/dali/core/format.h [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+#ifndef DALI_CORE_FORMAT_H_
+#define DALI_CORE_FORMAT_H_
+
+#include <string>
+
+namespace dali {
+
+template<class Delimiter>
+void print_delim(std::ostream&os, const Delimiter &delimiter) {
```
Comment (mzient) [inline include/dali/core/format.h]:
```suggestion
void print_delim(std::ostream &os, const Delimiter &delimiter) {
```

File: include/dali/core/format.h [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+ * In case there are no Args to operate on, return empty string.
+ */
+template <class Delimiter>
+std::string make_string_delim(const Delimiter &delimiter) {
+  return {};
+}
+
+
+/**
+ * Convenient version of @see make_string_delim, which takes a whitespace (' ')
```
Comment (mzient) [inline include/dali/core/format.h]:
```suggestion
 * @brief Prints args to a string, seperated by spaces.
```
"whitespace" is a more general term, whereas "space" is rather well defined (' ').

File: include/dali/core/format.h [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+template <class Delimiter, class T, class... Args>
+void print_delim(std::ostream &os, const Delimiter &delimiter, const T &val,
+                 const Args &... args) {
+  os << val << delimiter;
+  print_delim(os, delimiter, args...);
+}
+
+
+/**
+ * Creates std::string from arguments, as long as every element has `operator<<`
```
Comment (mzient) [inline include/dali/core/format.h]:
```suggestion
 * @brief Prints `args` to a string, separated by `delimiter`
```

File: include/dali/core/format.h [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+template <typename Delimiter>
+std::string make_string_delim(const Delimiter &) {
+  return {};
+}
+
+
+/**
+ * @brief Prints args to a string, seperated by spaces
+ */
+template <class... Args>
```
Comment (jantonguirao) [inline include/dali/core/format.h]: typename

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build
